File: Ruby/02.rb


# /*************************************
# *
# *	find the sum of the evem terms in the 
# *	Fibonacci sequence whose values do not exceed four million
# *
# **************************************/

class EvenFib
	require 'pry'

	attr_accessor :num_a, :num_b, :num_c, :sum

	def initialize(num_a = 0, num_b = 1, sum = 0)
		@num_a = num_a
		@num_b = num_b
		@num_c = @num_a + @num_b
		@sum = sum
	end

	# binding.pry
	
	def next_num
		@num_a = @num_b
		@num_b = @num_c
		@num_c = @num_a + @num_b
	end

	def is_even?(x)
		true if x % 2 == 0
	end

	def sum
		while @num_c < 4_000_000

			if is_even?(@num_c) 
				@sum += @num_c 
			end
			next_num
		end
		puts "#{@sum}"
	end
end

x = EvenFib.new
x.sum

File: python/solved/02.py

# /*************************************
# *
# *   find the sum of the evem terms in the 
# *   Fibonacci sequence whose values do not exceed four million
# *
# **************************************/




import time
from functools import reduce

start_time = time.time()

nums = [0,1]
while nums[len(nums)-1] <= 4000000:
    nums.append(nums[len(nums)-1] + nums[len(nums) - 2])

res = reduce(lambda x,y: x + y , list(filter(lambda x: x < 4000000 and x % 2 == 0, nums)))
print("answer: %s\n\n--- Complete in %s seconds ---" % (res, time.time() - start_time))

